File: resources/recipes/authentication/password-hashing/HashBcrypt.java
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

public class HashBcrypt {
    public static void main(String[] args) {
        String password = "supersecret";
        BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(12);

        long start = System.nanoTime();
        String hashed = encoder.encode(password);
        double hashMs = (System.nanoTime() - start) / 1_000_000.0;

        boolean ok = encoder.matches(password, hashed);

        System.out.printf("bcrypt hash: %s%n", hashed);
        System.out.printf("hash time: %.2f ms%n", hashMs);
        System.out.printf("verify: %b%n", ok);
    }
}

File: resources/recipes/authentication/password-hashing/hash_bcrypt.js
const bcrypt = require('bcrypt');

const password = 'supersecret';

async function main() {
  const start = Date.now();
  const hash = await bcrypt.hash(password, 12);
  const hashMs = Date.now() - start;
  console.log(`bcrypt hash: ${hash}`);
  console.log(`hash time: ${hashMs} ms`);
  const ok = await bcrypt.compare(password, hash);
  console.log(`verify: ${ok}`);
}

main().catch(console.error);
